databuilder/query_model_transforms.py

```diff
@@ -28,75 +42,184 @@ class PickOneRowPerPatientWithColumns(PickOneRowPerPatient):
-        # instead
-        selected_columns = frozenset(
-            SelectColumn(node.source, name) for name in column_names
+
+        # What columns are select from this patient frame?
+        selected_column_names = {
+            c.name for c in reverse_index[node] if isinstance(c, SelectColumn)
+        }
+
+        add_columns_to_pick(node, selected_column_names)
+        canonicalize_sort_and_filter_order(node)
+        add_extra_sorts(node, selected_column_names)
+
+
+def add_columns_to_pick(node, selected_column_names):
+    selected_columns = frozenset(
+        SelectColumn(node.source, c) for c in selected_column_names
+    )
+    force_setattr(node, "__class__", PickOneRowPerPatientWithColumns)
+    force_setattr(node, "selected_columns", selected_columns)
+
+
+def canonicalize_sort_and_filter_order(node):
+    table, filters, sorts = get_frame_operations(node.source)
+    source = table
+    for f in filters:
+        force_setattr(f, "source", source)
+        source = f
+    for s in sorts:
+        force_setattr(s, "source", source)
+        force_setattr(s.sort_by, "source", source)
+        source = s
+    force_setattr(node, "source", source)
+
+
+def add_extra_sorts(node, selected_column_names):
+    all_sorts, _ = pop_sorts(node.source)
+    # Add at the bottom of the stack
+    lowest_sort = all_sorts[0]
+
+    for column in calculate_sorts_to_add(all_sorts, selected_column_names):
+        new_sort = Sort(
+            source=lowest_sort.source,
+            sort_by=make_sortable(SelectColumn(lowest_sort.source, column)),
         )
-        # Create a new node which has the original attributes, plus the tuple of selected
-        # columns
-        new_node = PickOneRowPerPatientWithColumns(
-            source=node.source,
-            position=node.position,
-            selected_columns=selected_columns,
+        force_setattr(lowest_sort, "source", new_sort)
+        force_setattr(lowest_sort.sort_by, "source", new_sort)
+        lowest_sort = new_sort
+
+
+def calculate_sorts_to_add(all_sorts, selected_column_names):
+    # Don't duplicate existing direct sorts
+    direct_sorts = [
+        sort for sort in all_sorts if isinstance(sort.sort_by, SelectColumn)
+    ]
+    existing_sorted_column_names = {sort.sort_by.name for sort in direct_sorts}
+    sorts_to_add = selected_column_names - existing_sorted_column_names
+
+    # Arbitrary canonical ordering
+    return sorted(sorts_to_add)
+
+
+def make_sortable(col):
+    if get_series_type(col) == bool:
+        # Some databases can't sort booleans (including SQL Server), so we map them to integers
+        return Case(
+            cases={col: Value(2), Function.Not(col): Value(1)}, default=Value(0)
         )
-        # Update any references to the old node to point to the new one
-        for ref_node in references:
-            update_references(ref_node, node, new_node)
+    return col
```

Comment:
Do we need a test to demonstrate the new behaviour, that boolean values are sorted as we expect?

Reply:
We do! Added, thanks.